int32 indexing for Tensor Iterator Reduction

jjsjann123 opened the pull request:
Summary:
1. Enabling int32 indexing for cases where TI cannot accumulate in output due to
incompatible data types (e.g. Welford).
2. Updating Welford kernel to use int32 instead of int64 indexing on GPU.

This change improves performance for torch.var / torch.std

Implementation:
1. Allocated extra buffer to handle accumulation between sub Tensor Iterators.
2. Removed int64 indexing in gpu_reduce_kernel
3. WelfordOps now supports index type / combination typeas a template parameter.
While GPU uses int32_t and float, CPU implementation uses int64_t and double.

jjsjann123 committed:
int32 indexing for Tensor Iterator Reduction

Summary:
1. Enabling int32 indexing for cases where TI cannot accumulate in output due to
incompatible data types (e.g. Welford).
2. Updating Welford kernel to use int32 instead of int64 indexing on GPU.

This change improves performance for torch.var / torch.std

Implementation:
1. Allocated extra buffer to handle accumulation between sub Tensor Iterators.
2. Removed int64 indexing in gpu_reduce_kernel
3. WelfordOps now supports index type / combination typeas a template parameter.
While GPU uses int32_t and float, CPU implementation uses int64_t and double.

jjsjann123 commented:
Ping @ngimel @umanwizard @colesbury for visibility.
Benchmark scripts and perf number will follow shortly.

ngimel commented on aten/src/ATen/native/cuda/Reduce.cuh:
```diff
@@ -311,13 +314,29 @@ struct ReduceOp {
     auto out = (out_scalar_t*)((char*)dst + base_offsets[0]);
+    arg_t* acc = nullptr;
+    if (buf != nullptr) {
+      acc = (arg_t*)((char*)buf + (int)(base_offsets[0] * acc_buffer_multiplier));
```
how does this work with acc_buffer_multiplier being possibly non-integer?

jjsjann123 replied:
This should be fine: as `base_offsets[0]` is offset by bytes, given our data structure is assumed to be aligned, the end result would always be an integer.

ngimel replied:
I don't think it is, correct me if I'm wrong: base_offsets[0] will be promoted to float, multiplied with acc_buffer_multiplier, the result will be float and might be nonsensical. If result is larger than 2^24, not all integers are representable, and if you happen to hit one of those, then the result will be rounded.

jjsjann123 replied:
I stand corrected.

Luckily this is host code, so size_t should be perfectly safe. I can have a numerator/denominator.
To avoid overflow, I might need to reduce the fraction as well...

jjsjann123 replied:
This one is not host code :( But it's not in a loop. We should be good.

ngimel commented on aten/src/ATen/native/cuda/Reduce.cuh:
```diff
@@ -508,8 +538,38 @@ static void launch_reduce_kernel(const ReduceConfig& config, const R& reduction)
+struct AccumulationBuffer {
+  AccumulationBuffer() {}
+
+  AccumulationBuffer(size_t acc_stride, size_t out_stride, char* out_ptr, int64_t size) {
```
acc_stride and out_stride names are misleading, they are element sizes?

jjsjann123 replied:
they are sizeof(acc_t) & sizeof(out_t)
I'll rename them to acc_t_size, out_t_size.

ngimel commented on aten/src/ATen/native/cuda/Reduce.cuh:
```diff
@@ -508,8 +538,38 @@ static void launch_reduce_kernel(const ReduceConfig& config, const R& reduction)
+    if (size_factor_ == -1) {
+      return nullptr;
+    }
+    return acc_ptr_ + (int)((out_ptr - out_ptr_) * size_factor_);
```
I'm nervous here and in other places about pointer arithmetic involving floats. Also, I'm nervous about casting to (int) because it might conceivably overflow for large outputs and size_factors_.

jjsjann123 replied:
Fixed pointer arithmetic with casting and floats to properly use `size_t` instead.

jjsjann123 commented on aten/src/ATen/native/SharedReduceOps.h:
```diff
@@ -31,50 +31,53 @@
   inline C10_DEVICE acc_t reduce(acc_t acc, scalar_t data) const {
     acc_scalar_t delta = data - acc.mean;
-    acc_scalar_t new_mean = acc.mean + delta / (acc.n + 1);
+    acc_scalar_t new_mean = acc.mean + delta / (acc.nf + 1);
```
`acc.nf` in `reduce` is only used for the division and gets updated from `acc.n` (`int32_t`) for each increament. So this should be fine.
I'll put a comment here.

jjsjann123 commented on aten/src/ATen/native/SharedReduceOps.h:
```diff
@@ -31,50 +31,53 @@
       acc.m2 + delta * new_delta,
-      acc.n + 1
+      acc.n + 1,
+      combine_t(acc.n + 1),
```
Here's what I was talking about. I was counting using `acc.n` and update `acc.nf` with it here `combine_t(acc.n+1)`

jjsjann123 committed:
Addressing review comment.

Pointer arithmetic with floats and casting has been updated to safely use size_t

jjsjann123 committed:
updated buffer allocation to handle non-contiguous output tensor

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into TI_subiterator_PR

jjsjann123 commented:
Test failure seems to be machine error.

jjsjann123 commented:
![image](https://user-images.githubusercontent.com/3709243/53521490-1d6d1a80-3a8d-11e9-8623-711c4042b7fd.png)

Benchmark torch.std
Red shows regression and green is speedup.

Benchmark script:
```
import torch
nrep = 100


def bench(size, fn):
   x=torch.ones([size], device='cuda', dtype = torch.float).view(-1, 512)
   torch.cuda.synchronize()
   import time
   start = time.time()
   for i in range(nrep):
      fn(x,0)
   torch.cuda.synchronize()
   end = time.time()
   return ((end-start)/nrep)


stdbws = []
sizes = []
size = 102400
while size < 1024000000:
   time = bench(size, torch.std)
   bw = size*4/time*1e-9
   stdbws.append(bw)
   sizes.append(size)
   size *= 2
for size, stdbw in zip(sizes, stdbws):
   print(size, stdbw)
```

jjsjann123 commented:
![image](https://user-images.githubusercontent.com/3709243/53523318-5d360100-3a91-11e9-9a3a-48a724c2b769.png)

I have another PR that resolves the dependent loop unrolling in current Tensor Iterator. It gives further speedup (table above shows relative speedup to int32 indexing)
Perf fluctuates between runs (especially with smaller sizes). But I do see noticeably better perf and proper sass instructions.

The other PR is dependent on this one, I'll issue the other PR once this is merged.

umanwizard commented on aten/src/ATen/native/cuda/Reduce.cuh:
```diff
@@ -518,15 +601,37 @@ inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t id
+    }
+  }
+
+  // if (can_accumulate_in_output && !can_use_32bit_indexing) {
```
let's remove the commented-out code

jjsjann123 replied:
Oops

umanwizard commented on aten/src/ATen/native/cuda/Reduce.cuh:
```diff
@@ -508,8 +555,44 @@ static void launch_reduce_kernel(const ReduceConfig& config, const R& reduction)
 template <typename scalar_t, typename out_scalar_t, typename ops_t, typename ident_t=double>
-inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t ident=0) {
+inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t ident=0,
+                              std::shared_ptr<AccumulationBuffer> acc_buf_ptr={}) {
```
Does `acc_buf_ptr` really need to be reference-counted/shared?

Why can't we just pass around a raw `AccumulationBuffer *` pointer, and allocate the `AccumulationBuffer` on the stack in the top-level call (so it will be properly deallocated when that call returns)

jjsjann123 replied:
Fair enough, I'll switch to unique_ptr and pass raw pointer instead.

jjsjann123 commented on aten/src/ATen/native/cuda/Reduce.cuh:
```diff
@@ -35,6 +35,17 @@ static inline int last_pow2(int n) {
 }
 
+// returns reduced fraction numerator & denominator
+C10_HOST_DEVICE static void reduce_fraction(size_t &numerator, size_t &denominator) {
```
This looks better.
Back then, I thought num / denom are only sizeof(data_t), so didn't bother search for a better way. 🤣

jjsjann123 committed:
Addressing review comments

updating fraction reduction function;
change shared_ptr to unique_ptr for recursive gpu_reduce_kernel call;

apaszke commented on aten/src/ATen/native/cuda/Reduce.cuh:
```diff
@@ -35,6 +35,25 @@ static inline int last_pow2(int n) {
+      // swap(a,b)
+      a ^= b;
+      b ^= a;
+      a ^= b;
```
IIRC XOR swaps are not really the fastest way to do this. Can't you just `std::swap` which is so much more readable?

apaszke replied:
See [the commentary on the Wikipedia article](https://en.wikipedia.org/wiki/XOR_swap_algorithm#Reasons_for_avoidance_in_practice).

jjsjann123 replied:
IIRC, there's no equivalent device function for std::swap.

For the equivalent issue, I could update the condition check earlier to ensure that this function is safe (also proper comment as well)

apaszke replied:
I thought `std` works in device functions, doesn't it? Otherwise, let's just use a temporary.

apaszke commented on aten/src/ATen/native/cuda/Reduce.cuh:
```diff
@@ -593,42 +707,24 @@ inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t id
-
-    launch_reduce_kernel<ReduceConfig::MAX_NUM_THREADS>(config, reduce);
-  }
+  assert(can_use_32bit_indexing);
```
Please don't use `assert`. `AT_ASSERT` is what you want.

jjsjann123 replied:
Will update

umanwizard commented:
Hi @jjsjann123 , could you address @apaszke 's comments, and also rebase to see if the failing tests pass then?

jjsjann123 commented on aten/src/ATen/native/cuda/Reduce.cuh:
```diff
@@ -508,8 +563,44 @@ static void launch_reduce_kernel(const ReduceConfig& config, const R& reduction)
+
+  AccumulationBuffer(size_t acc_t_size, size_t out_t_size, char* out_ptr, int64_t size) {
+    out_ptr_ = (char*)out_ptr;
+    if (out_t_size > acc_t_size) {
```
Update this to `out_t_size >= acc_t_size`

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into TI_subiterator_PR

jjsjann123 committed:
adressing comment issues

jjsjann123 commented:
Failure looks like infra issue. (Can I) How do I trigger circleci test?
